src/Providers/OpenAiCompatibleImplementation/AbstractOpenAiCompatibleTextGenerationModel.php

```diff
@@ -0,0 +1,727 @@
+                        'Unsupported output modality "%s".',
+                        $modality
+                    )
+                );
+            }
+        }
+        return $prepared;
+    }
+
+    /**
+     * Prepares the tools parameter for the API request.
+     *
+     * @since n.e.x.t
+     *
+     * @param list<FunctionDeclaration> $functionDeclarations The function declarations.
+     * @return list<array<string, mixed>> The prepared tools parameter.
+     */
+    protected function prepareToolsParam(array $functionDeclarations): array
+    {
+        $tools = [];
+        foreach ($functionDeclarations as $functionDeclaration) {
+            $tools[] = [
+                'type' => 'function',
+                'function' => $functionDeclaration->toArray(),
+            ];
+        }
+
+        return $tools;
+    }
+
+    /**
+     * Prepares the response format parameter for the API request.
+     *
+     * This is only called if the output MIME type is `application/json`.
+     *
+     * @since n.e.x.t
+     *
+     * @param array<string, mixed>|null $outputSchema The output schema.
+     * @return array<string, mixed> The prepared response format parameter.
+     */
+    protected function prepareResponseFormatParam(?array $outputSchema): array
+    {
+        if (is_array($outputSchema)) {
+            return [
+                'type' => 'json_schema',
+                'json_schema' => $outputSchema,
+            ];
+        }
+
+        return [
+            'type' => 'json_object',
+        ];
+    }
+
+    /**
+     * Creates a request object for the provider's API.
+     *
+     * @since n.e.x.t
+     *
+     * @param HttpMethodEnum $method The HTTP method.
+     * @param string $path The API endpoint path, relative to the base URI.
+     * @param array<string, string|list<string>> $headers The request headers.
+     * @param string|array<string, mixed>|null $data The request data.
+     * @return Request The request object.
+     */
+    abstract protected function createRequest(
+        HttpMethodEnum $method,
+        string $path,
+        array $headers = [],
+        $data = null
+    ): Request;
+
+    /**
+     * Throws an exception if the response is not successful.
+     *
+     * @since n.e.x.t
+     *
+     * @param Response $response The HTTP response to check.
+     * @throws ResponseException If the response is not successful.
+     */
+    protected function throwIfNotSuccessful(Response $response): void
+    {
+        ResponseUtil::throwIfNotSuccessful($response);
+    }
+
+    /**
+     * Parses the response from the API endpoint to a generative AI result.
+     *
+     * @since n.e.x.t
+     *
+     * @param Response $response The response from the API endpoint.
+     * @return GenerativeAiResult The parsed generative AI result.
+     */
+    protected function parseResponseToGenerativeAiResult(Response $response): GenerativeAiResult
+    {
+        $responseData = $response->getData();
```

Comment:
Let's define a response shape type for the class to be used throughout, as there are a number of explicit `@var` instances that would be cleaned up by that.

Note that we can define multiple types and combine them for internal arrays of arrays like tools.